### DISCUSSION TITLE
Is git LFS suitable for large folder consist of multiple small files?

### DISCUSSION THREAD

**Discussion by Jiazheng411** [git-lfs/git-lfs#4676] [General]
I have a resource folder around 2G, but each file inside is only 2-3 MB usually, maximum 11 MB, total 1000+ files, is this a suitable use case for git LFS? If not, what do you recommend to manage the files?

**Comment by chrisd8088** [comment]
I would always suggest experimenting a little to see if Git LFS can be accommodated in your workflow, but in terms of the files themselves, this should not be a problem, I would think.  Git itself handles looking through the directories; it then calls Git LFS on each file which matches whatever rule you set up in `.gitattributes` -- in your case, it might be "all files in this resource folder".  So Git LFS doesn't traverse the hierarchy on its own; that's Git's job, and it should handle large directories just fine.  As for the file sizes, those seem like good candidates for Git LFS.

**Reply by tedsecretsource** [reply]
Thanks for much for the feedback and I will experiment. However, I have a similar question. I'm working on a computer vision machine learning project. It has 10s of thousands of image files between 100KB and 10MB. The vast majority (80%) are under 256KB. Because of the name, Large File Support, I'm inclined to think this isn't really the right tool to help manage the ballooning repo size. Does LFS help keep the repo size manageable (under 5GB, for example) in this situation?

Many thanks in advance!

Ted

## References
https://stackoverflow.com/questions/35787080/how-well-does-git-lfs-handle-small-files
https://support.atlassian.com/bitbucket-cloud/docs/use-git-lfs-with-bitbucket/
https://medium.com/swlh/learning-about-git-large-file-system-lfs-72e0c86cfbaf

**Reply by derekperkins** [reply]
@tedsecretsource we're in a similar situation to you, where most of our files are in the 100-300kb range. What conclusion did you come to?

**Reply by tedsecretsource** [reply]
@derekperkins Google Drive! I lament not having the images be part of the repo but even with LFS it's too cumbersome.

Another alternative is a self-hosted remote git repository that doesn't have the 5GB limitation but if I did go down that route, I would probably implement LFS just to keep duplications and repo size to a minimum (so yes, it will help but for my needs Google Drive is adequate).

**Reply by derekperkins** [reply]
Thanks for the quick response! I definitely wasn't expecting Google Drive as the answer, but that's engineering for you - pick what works. :)

**Reply by ReHoss** [reply]
@tedsecretsource , why not the compressed folder solution (.zip or .tar.gz)?

**Reply by tedsecretsource** [reply]
@ReHoss because we were constantly changing individual images to see the effects of updating the training set. Keeping all the images in a single archive is problematic. We've learned that bitbucket simply isn't the best tool for what we were doing. Thanks for the suggestion in any case.

**Reply by ReHoss** [reply]
> @ReHoss because we were constantly changing individual images to see the effects of updating the training set. Keeping all the images in a single archive is problematic. We've learned that bitbucket simply isn't the best tool for what we were doing. Thanks for the suggestion in any case.

@tedsecretsource  To me, the training set filtering/transformation logic should be handled by a post-processing script or function starting from the initial raw `.zip` file.
The post-processing part would implement your different training set customisation choices.

**Comment by chrisd8088** [comment]
My first thought is that you might want to keep those kinds of files in another system, since Git is really focussed on source code management, not binary file management.  There are lots of object storage and file management systems which might be useful.

If you do want to keep them in Git, then as a broad general statement Git LFS will help, because despite the name what it does is store files of any size matching a pattern (like `*.jpg`) outside of the Git repo.  But it's important to note that the main place this helps is in the size of the repo *on the remote*.  That is to say, if you're pushing your repo to a forge like GitLab or GitHub, they won't see or store copies of your image files in the repo they store; the image files will be pushed and pulled to another system like Amazon S3.  Your Git client will retrieve them from there and update them there as well.

But if you have all those files in the `HEAD` of your repo, Git LFS may well double the size of the storage on your local disk.  That's because, when you check out a branch, the Git LFS client will "smudge" any small Git LFS pointer files it finds in the working tree and expand them into the full files.  To do that it fetches the real file from the external storage (say, Amazon S3) and places a copy under `.git/lfs/objects`.  Then it replaces the pointer file in the repo's working copy with the real file, so that's what you see as you work in your normal work tree.

If your image files are divided up across many branches you may be able to keep this local size down using `git lfs prune` and pruning everything that's not in your immediate working branch.  And you can also check out without smudging using `GIT_LFS_SKIP_SMUDGE=1`, but then you don't have any of your image files in your working copy.

So it very much depends on your workflow, your repo shape, etc.  But if you just have a big pile of images that you need to keep somewhere, Git may not be the right place in the first place, and you might want to look at other options.  (Among other things, Git wants to `diff` things across revisions, and binaries don't fit that model.  Git really is a source code management tool, not a generic file management tool.)

**Reply by tedsecretsource** [reply]
Thank you so much for the response and clarifications. Indeed, I probably would benefit from another system. FWIW, the reason I went with git is because I'm learning about Machine Learning and wanted to see what the effects of one small change would have on the training the model and git seemed like a good choice because of its granularity. I still think it is, but definitely need to store the _actual_ images elsewhere using LFS as you describe. Thanks again!
